### PR TITLE
Redo header stamping

### PR DESCRIPTION
**Pull request by dobosevych**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

### CODE REVIEW

File: t/unit/tasks/test_canvas.py [hunk 1]
```diff
@@ -1240,6 +1249,29 @@ def test_apply_contains_chords_containing_empty_chord(self):
...
+        sig_1_res = sig_1.freeze()
+        sig_2_res = sig_2.freeze()
+
+        g = chord([sig_1, sig_2], self.xsum.s(), app=self.app)
+        g_res = g.freeze()
+        g.stamp(groups=g_res.id)
+        g.apply()
+
+        assert sig_1_res._get_task_meta()['groups'] == [g_res.id]
+        assert sig_2_res._get_task_meta()['groups'] == [g_res.id]
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:1274]: We need to assert that xsum doesn't belong to any group.

File: celery/canvas.py [hunk 2]
```diff
@@ -1490,6 +1542,21 @@ def freeze(self, _id=None, group_id=None, chord=None,
         self.id = self.tasks.id
         return body_result
 
+    def stamp(self, visitor=None, **headers):
+        if visitor:
+            headers.update(visitor.on_group_start(self, **headers))
```
Comment (thedrow) [inline celery/canvas.py:1547]: This is the wrong thing to do here.
We discussed having an `on_chord_header_start`/`on_chord_header_end` which by default calls `on_group_start` but can be overridden if needed.
This implementation also overlooks the chord's body completely.

File: celery/canvas.py [hunk 3]
```diff
@@ -1292,7 +1341,10 @@ def _apply_tasks(self, tasks, producer=None, app=None, p=None,
     def _freeze_gid(self, options):
         # remove task_id and use that as the group_id,
         # if we don't remove it then every task will have the same id...
-        options = dict(self.options, **options)
+        options = dict(self.options, **{
+                k: v for k, v in options.items()
+                if k not in self._IMMUTABLE_OPTIONS or k not in self.options
+            })
```
Comment (thedrow) [inline celery/canvas.py:1347]: The modern form for merging dictionaries is `{**d1, **d2}`.

File: celery/canvas.py [hunk 4]
```diff
@@ -1184,24 +1229,28 @@ def apply(self, args=None, kwargs=None, **options):
         if not self.tasks:
             return self.freeze()  # empty group returns GroupResult
         options, group_id, root_id = self._freeze_gid(options)
+        #self.stamp(visitor=OptionsStamping(), **options)
+        #group_id = self.options['group_id']
+        #root_id = options.get('root_id')
```
Comment (thedrow) [inline celery/canvas.py:1234]: Why is this commented out?

**Review by thedrow — CHANGES_REQUESTED** [review]

File: t/unit/tasks/test_canvas.py (hunk 1 above)
Comment (dobosevych) [reply t/unit/tasks/test_canvas.py:1278]: added

File: celery/canvas.py (hunk 2 above)
Comment (dobosevych) [reply celery/canvas.py]: added + on_chord_body

File: t/unit/tasks/test_canvas.py [hunk 5]
```diff
@@ -1240,6 +1235,31 @@ def test_apply_contains_chords_containing_empty_chord(self):
...
+        sig_1_res = sig_1.freeze()
+        sig_2_res = sig_2.freeze()
+        sig_sum = self.xsum.s()
+        sig_sum.freeze()
+
+        g = chord([sig_1, sig_2], sig_sum, app=self.app)
+        g.apply()
+
+        with pytest.raises(AttributeError):
+            sig_sum._get_task_meta()['groups']
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:1260]: You need the AsyncResult returned from freeze to do that

File: t/unit/tasks/test_canvas.py [hunk 6]
```diff
@@ -1240,6 +1235,31 @@ def test_apply_contains_chords_containing_empty_chord(self):
...
+        sig_2 = self.add.s(4, 4)
+        sig_1_res = sig_1.freeze()
+        sig_2_res = sig_2.freeze()
+        sig_sum = self.xsum.s()
+        sig_sum.freeze()
+
+        g = chord([sig_1, sig_2], sig_sum, app=self.app)
+        g.apply()
+
+        with pytest.raises(AttributeError):
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:1259]: This should raise a KeyError instead.

File: t/unit/tasks/test_canvas.py [hunk 7]
```diff
@@ -696,6 +710,28 @@ def test_group_stamping_two_levels(self):
...
+        sig_2 = self.add.s(2, 2) | self.replaced.s(8)
+        sig_1_res = sig_1.freeze()
+        sig_2_res = sig_2.freeze()
+
+        g = group(sig_1, sig_2, app=self.app)
+        g_res = g.freeze()
+        g.apply()
+
+        assert sig_1_res._get_task_meta()['groups'] == [g_res.id]
+        assert sig_2_res._get_task_meta()['groups'] == [g_res.id]
```
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:733]: Are we sure this checks the result of `replaced`?

Comment (dobosevych) [reply t/unit/tasks/test_canvas.py:733]: sig_2_res should be stamped with g_res.id
sig_2_res would without stamping have an empty groups list, as this is result of replaced application

**Review by auvipy — CHANGES_REQUESTED** [review]
I would love to see more test coverage & possibly some integration test as well for this

**Comment by dobosevych** [comment]
> 

I improved coverage, working on integration tests.

**Comment by auvipy** [comment]
OK cool great

**Comment by auvipy** [comment]
any related issues those can be closed after merging this?

File: celery/canvas.py [hunk 8]
```diff
@@ -77,32 +78,124 @@ def _merge_dictionaries(d1, d2):
...
+    """
+    def __init__(self, groups=None):
+        self.groups = groups or []
+
+    def on_group_start(self, group, **headers) -> dict:
+        if group.id is None:
+            group.freeze()
+        if group.id not in self.groups:
+            self.groups.append(group.id)
         return {'groups': list(self.groups)}
```
Comment (auvipy) [inline celery/canvas.py:188]:
```suggestion
        return {'groups': [self.groups]}
```

**Review by auvipy — COMMENTED** [review]
what about using this?

File: celery/canvas.py (hunk 8 above)
Comment (dobosevych) [reply celery/canvas.py:188]:
```
>>> lst = [1, 2]
>>> [lst] # is a list of lists
>>> list(lst) # is a copy of list, we do that to not work with reference
```

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
lets move to the another open PR

**Pull request merged**